### PR TITLE
Websocket support

### PR DESCRIPTION
**Pull request by ekoby**
This PR is in draft. Thing still left TODO:

- [x] handle websocket pings
- [x] check for memory leaks
- [x] graceful close
- [ ] fix mac build

### CODE REVIEW

File: include/uv_mbed/um_websocket.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2019 Netfoundry, Inc.
```
Comment (smilindave26) [inline include/uv_mbed/um_websocket.h:2]: 2020

File: sample/ws-client.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2019 Netfoundry, Inc.
```
Comment (smilindave26) [inline sample/ws-client.c:2]: 2020

File: src/http.c [hunk 3]
```diff
@@ -481,53 +290,7 @@ static void process_requests(uv_async_t *ar) {
...
-                                    "%s: %s\r\n", h->name, h->value);
-            }
-
-            if (need_host) {
-                req.len += snprintf(req.base + req.len, 8196 - req.len,
-                                    "Host: %s\r\n", c->host);
-            }
-            req.len += snprintf(req.base + req.len, 8196 - req.len,
-                                "\r\n");
+            req.len = http_req_write(c->active, req.base, 8196);
```
Comment (smilindave26) [inline src/http.c:293]: good ol' 8196.

File: src/http_req.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    if (req->resp_cb != NULL) {
+        req->resp_cb(&req->resp, req->data);
+    }
+    return 0;
+}
+
+static int http_header_field_cb(http_parser *parser, const char *f, size_t len) {
+    um_http_req_t *req = parser->data;
+    if (req->resp.headers == NULL) {
+        req->resp.headers = calloc(20, sizeof(um_http_hdr));
```
Comment (smilindave26) [inline src/http_req.c:163]: is this realloc'd somewhere if over 20 (or errored out at least)?

File: src/http_req.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//
+// Created by eugene on 6/16/20.
```
Comment (smilindave26) [inline src/http_req.h:18]: :)

File: src/tls_link.c [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    cb(source, 0, arg);
+
+    return rc;
+}
+
+static void tls_close(uv_link_t *l, uv_link_t *source, uv_link_close_cb close_cb) {
+    UM_LOG(VERB, "closing TLS link");
+//    um_http_t *clt = l->data;
+//
+//    clt->tls->api->free_engine(clt->engine);
```
Comment (smilindave26) [inline src/tls_link.c:157]: ? not needed or is there a follow-up needed?

Comment (ekoby) [reply src/tls_link.c:157]: not needed client is in charge of the engine lifecycle

File: src/http_req.c (hunk 4 above)
Comment (ekoby) [reply src/http_req.c]: changed response headers into `um_headers_list`

**Pull request merged**